player/command.c

```diff
@@ -2360,7 +2375,13 @@ static int mp_property_vo_imgparams(void *ctx, struct m_property *prop,
     if (valid != M_PROPERTY_VALID)
         return valid;
 
-    return property_imgparams(get_video_out_params(ctx), action, arg);
+    struct mp_image_params p = get_video_out_params(ctx);
+
+    MPContext *mpctx = ctx;
+    if (mpctx->video_out)
+        vo_control(mpctx->video_out, VOCTRL_HDR_METADATA, &p.color.hdr);
```

Comment:
Since you added `pl_hdr_metadata` directly to `mp_colorspace` earlier, you could remove this entire mechanism and just read the params directly. I suppose gpu_next would need to be changed so it saves the `pl_hdr_metadata` directly in the image params.

Reply:
This is the only way to extract peak detection values from gpu-next. See `pl_renderer_get_hdr_metadata` call in `vo_gpu_next`.

Reply:
Can't you just put the `pl_hdr_metadata` in `vo_gpu_next` in `frame->current->params` in the draw_frame call?

Reply:
The kicker is that `mp_property_vo_imgparams` does not read params from `vo_get_current_frame`, but from `mpctx->vo_chain->filter->output_params`

I can look what can be done about it. Indeed probably we can simplify it. But not sure it is safe to mutate current frame.

Reply:
I did not verify, but I am pretty sure that will be a copy of whatever `vo_get_current_frame` is plus possibly some other things (like filters or whatever). So as long as the image params in the actual vo are modified, it should hopefully just work.

Reply:
I agree, conceptually this is probably correct way to do that. And if anything is not working it should be fixed there instead.

Reply:
I've tried quickly this, but it doesn't work. After looking at `vo.c`, current frame after draw is not longer current, so the "overridden" params are lost.

``` diff
diff --git a/player/command.c b/player/command.c
index 39d7a659f4..533a1279cd 100644
--- a/player/command.c
+++ b/player/command.c
@@ -2368,17 +2368,16 @@ static struct mp_image_params get_video_out_params(struct MPContext *mpctx)
 static int mp_property_vo_imgparams(void *ctx, struct m_property *prop,
                                     int action, void *arg)
 {
+    MPContext *mpctx = ctx;
+    if (!mpctx->video_out)
+        return M_PROPERTY_UNAVAILABLE;
+
     int valid = m_property_read_sub_validate(ctx, prop, action, arg);
     if (valid != M_PROPERTY_VALID)
         return valid;
 
-    struct mp_image_params p = get_video_out_params(ctx);
-
-    MPContext *mpctx = ctx;
-    if (mpctx->video_out)
-        vo_control(mpctx->video_out, VOCTRL_HDR_METADATA, &p.color.hdr);
-
-    return property_imgparams(p, action, arg);
+    const struct mp_image *mpi = vo_get_current_frame(mpctx->video_out);
+    return property_imgparams(mpi->params, action, arg);
 }
 
 static int mp_property_dec_imgparams(void *ctx, struct m_property *prop,
diff --git a/video/out/vo.h b/video/out/vo.h
index 6b832b28e8..18291bd0d0 100644
--- a/video/out/vo.h
+++ b/video/out/vo.h
@@ -122,8 +122,6 @@ enum mp_voctrl {
 
     /* private to vo_gpu and vo_gpu_next */
     VOCTRL_EXTERNAL_RESIZE,
-
-    VOCTRL_HDR_METADATA,            // struct pl_hdr_metadata*
 };
 
 // Helper to expose what kind of content is currently playing to the VO.
diff --git a/video/out/vo_gpu_next.c b/video/out/vo_gpu_next.c
index d8ea1c09d4..e98dc891a0 100644
--- a/video/out/vo_gpu_next.c
+++ b/video/out/vo_gpu_next.c
@@ -1078,13 +1078,11 @@ static void draw_frame(struct vo *vo, struct vo_frame *frame)
     }
 
     if (cur_frame) {
-        p->last_hdr_metadata = cur_frame->color.hdr;
+        frame->current->params.color = cur_frame->color;
 #if PL_API_VER >= 314
         // Augment metadata with peak detection max_pq_y / avg_pq_y
-        pl_renderer_get_hdr_metadata(p->rr, &p->last_hdr_metadata);
+        pl_renderer_get_hdr_metadata(p->rr, &frame->current->params.color.hdr);
 #endif
-    } else {
-        p->last_hdr_metadata = (struct pl_hdr_metadata){0};
     }
 
     p->is_interpolated = mix.num_frames > 1;
@@ -1472,10 +1470,6 @@ static int control(struct vo *vo, uint32_t request, void *data)
         return true;
     }
 
-    case VOCTRL_HDR_METADATA:
-        *(struct pl_hdr_metadata *) data = p->last_hdr_metadata;
-        return true;
-
     case VOCTRL_SCREENSHOT:
         video_screenshot(vo, data);
         return true;

```

Reply:
I don't know. I guess we can clean this up next time...

Reply:
Did you try it while leaving the `get_video_out_params(ctx);` part the same but with the other changes?

Reply:
`get_video_out_params` doesn't include those either. VO drivers doesn't have any back channel to send data back. Same as with `VOCTRL_PERFORMANCE_DATA` they keep "last" data internaly.

Reply:
Hmm disappointing. I guess something for later then.